lib/oplog/oplog_v2_converter.go

```diff
@@ -0,0 +1,193 @@
+	}
+}
+
+// Given a map, with other maps possibly nested under it, returns the
+// flattened object keys. E.g.:
+//
+//	{
+//	  a: {
+//	    b: {
+//	      c: [{d: 1}],
+//	      e: 2
+//	    },
+//	    f: 3
+//	  }
+//	}
+//
+// becomes
+// ['a.b.c', 'a.b.e', 'a.f']
+func flatObjectKeys(prefix string, obj map[string]interface{}) []string {
+	acc := []string{}
+
+	for key, val := range obj {
+		switch typedVal := val.(type) {
+		case map[string]interface{}:
+			if len(typedVal) == 0 {
+				acc = append(acc, prefix+key)
+			} else {
+				acc = append(acc, flatObjectKeys(prefix+key+".", typedVal)...)
+			}
+		default:
+			acc = append(acc, prefix+key)
+		}
+	}
+
+	return acc
+}
+
+func getChangedFieldsFromOplogV2UpdateDeep(diffMap map[string]interface{}, prefix string) []string {
+	fields := []string{}
+
+	for operationKey, operation := range diffMap {
+		if operationKey == "i" || operationKey == "u" || operationKey == "d" {
+			// indicates an insert, update, or delete of a whole subtree
+			operationMap, operationMapOK := operation.(map[string]interface{})
+			if !operationMapOK {
+				metricUnprocessableChangedFields.Inc()
+				log.Log.Errorw("Oplog data for non-replacement v2 update contained a i/u/d key with a non-map value",
+					"op", diffMap)
+				continue
+			}
+
+			fields = append(fields, flatObjectKeys(prefix, operationMap)...)
+		} else if isArrayOperator(operation) {
+			operationMap, operationMapOK := operation.(map[string]interface{})
+			if !operationMapOK {
+				metricUnprocessableChangedFields.Inc()
+				log.Log.Errorw("Oplog data for non-replacement v2 update contained an array operator key with a non-map value",
+					"op", diffMap)
+				continue
+			}
+
+			for arrayOperatorKey := range operationMap {
+				if arrayOperatorKey == "a" {
+					continue
+				}
+
+				fields = append(fields, prefix+operationKey[1:]+"."+arrayOperatorKey[1:])
+			}
+		} else if strings.HasPrefix(operationKey, "s") {
+			// indicates an insert, update, or delete of a whole subtree
+			operationMap, operationMapOK := operation.(map[string]interface{})
+			if !operationMapOK {
+				metricUnprocessableChangedFields.Inc()
+				log.Log.Errorw("Oplog data for non-replacement v2 update contained a s-field key with a non-map value",
+					"op", diffMap)
+				continue
+			}
+
+			// indicates a sub-field set
+			fields = append(fields, getChangedFieldsFromOplogV2UpdateDeep(operationMap, prefix+operationKey[1:]+".")...)
+		} else if operationKey == "a" {
+			// ignore
+			continue
+		} else {
+			metricUnprocessableChangedFields.Inc()
+			log.Log.Errorw("Oplog data for non-replacement v2 update contained a field that was not an i/u/d or an s-prefixed field",
+				"op", diffMap)
+			continue
+		}
+
+	}
+
+	return fields
+}
+
+func getChangedFieldsFromOplogV2UpdateShallow(diffMap map[string]interface{}) []string {
+	fields := []string{}
+	for operationKey, operation := range diffMap {
+		if operationKey == "i" || operationKey == "u" || operationKey == "d" {
```

Comment:
Does this need to be repeated twice between shallow and deep? Are these sections likely to diverge too fast to be worth putting into a separate function?

Reply:
I think this is fine, I don't think it'll diverge too fast. I don't expect the v2 to change until v3 comes out, if ever.